parlai/agents/starspace/starspace.py

```diff
@@ -0,0 +1,459 @@
+                           default=None, type=str,
+                           help='File of cands to use for prediction')
+
+    def __init__(self, opt, shared=None):
+        """Set up model if shared params not set, otherwise no work to do."""
+        super().__init__(opt, shared)
+        opt = self.opt
+        self.reset_metrics()
+        self.id = 'Starspace'
+        self.NULL_IDX = 0
+        self.cands = torch.LongTensor(1, 1, 1)
+        self.ys_cache = []
+        self.ys_cache_sz = opt['cache_size']
+        self.truncate = opt['truncate'] if opt['truncate'] > 0 else None
+        self.history = {}
+        if shared:
+            self.threadindex = shared['threadindex']
+            print("[ creating Starspace thread " + str(self.threadindex)  + " ]")
+            # set up shared properties
+            self.dict = shared['dict']
+            self.model = shared['model'] #Starspace(opt, len(self.dict))
+        else:
+            print("[ creating StarspaceAgent ]")
+            # this is not a shared instance of this class, so do full init
+            if opt['dict_file'] is None and opt.get('model_file'):
+                # set default dict-file if not set
+                opt['dict_file'] = opt['model_file'] + '.dict'
+            # load dictionary and basic tokens & vectors
+            self.dict = DictionaryAgent(opt)
+
+            self.model = Starspace(opt, len(self.dict), self.dict)
+            if opt.get('model_file') and os.path.isfile(opt['model_file']):
+                self.load(opt['model_file'])
+            self.model.share_memory()
+
+        # set up modules
+        self.criterion = torch.nn.CosineEmbeddingLoss(margin=opt['margin'], size_average=False)
+        self.reset()
+        self.fixedCands = False
+        if self.opt.get('fixed-candidates-file'):
+            self.fixedCands = load_cands(self.opt.get('fixed-candidates-file'))
+
+    def reset(self):
+        """Reset observation and episode_done."""
+        self.observation = None
+        self.episode_done = True
+        # set up optimizer
+        lr = self.opt['learningrate']
+        optim_class = StarspaceAgent.OPTIM_OPTS[self.opt['optimizer']]
+        kwargs = {'lr': lr}
+        self.optimizer = optim_class(self.model.parameters(), **kwargs)
+
+    def share(self):
+        """Share internal states between parent and child instances."""
+        shared = super().share()
+        shared['dict'] = self.dict
+        shared['model'] = self.model
+        return shared
+
+    def override_opt(self, new_opt):
+        """Set overridable opts from loaded opt file.
+
+        Print out each added key and each overriden key.
+        Only override args specific to the model.
+        """
+        model_args = {'hiddensize', 'embeddingsize', 'numlayers', 'optimizer',
+                      'encoder', 'decoder', 'lookuptable', 'attention',
+                      'attention_length'}
```

Comment:
if you're keeping this may want to update these. not sure if you need anything other than `embeddingsize` (and `tfidf`?)

Reply:
we need to overhaul this whole system. will talk to you

Reply:
(but not for this diff)